bigtable/internal/testproxy/proxy.go

```diff
@@ -0,0 +1,976 @@
+		if c == lim {
+			return false
+		}
+		rpb, err := rowToProto(r)
+		if err != nil {
+			return false
+		}
+		rowsPb = append(rowsPb, rpb)
+		return true
+	})
+	s.clientsLock.Unlock()
+
+	res := &pb.RowsResult{
+		Status: &status.Status{
+			Code: int32(codes.OK),
+		},
+		Row: rowsPb,
+	}
+
+	return res, nil
+}
+
+// MutateRow responds to the MutateRow RPC. This methods applies a series of
+// changes (or deletions) to a single row in a table.
+func (s *goTestProxyServer) MutateRow(ctx context.Context, req *pb.MutateRowRequest) (*pb.MutateRowResult, error) {
+	s.clientsLock.Lock()
+	btc, exists := s.clientIDs[req.ClientId]
+	if !exists {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+
+	rrq := req.GetRequest()
+	if rrq == nil {
+		return nil, stat.Error(codes.InvalidArgument, "request to MutateRow() is missing inner request")
+	}
+
+	mPbs := rrq.Mutations
+	m := mutationFromProto(mPbs)
+
+	t := btc.c.Open(rrq.TableName)
+	row := rrq.RowKey
+
+	ctx = btc.addCancelFunction(ctx)
+
+	err := t.Apply(ctx, string(row), m)
+	if err != nil {
+		return nil, err
+	}
+	s.clientsLock.Unlock()
+
+	res := &pb.MutateRowResult{
+		Status: &status.Status{
+			Code: int32(codes.OK),
+		},
+	}
+	return res, nil
+}
+
+// BulkMutateRows responds to the BulkMutateRows RPC. This method applies a
+// series of changes or deletions to multiple rows in a single call.
+func (s *goTestProxyServer) BulkMutateRows(ctx context.Context, req *pb.MutateRowsRequest) (*pb.MutateRowsResult, error) {
+	s.clientsLock.Lock()
+	btc, exists := s.clientIDs[req.ClientId]
+	if !exists {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+
+	rrq := req.GetRequest()
+	if rrq == nil {
+		return nil, stat.Error(codes.InvalidArgument, "request to BulkMutateRows() is missing inner request")
+	}
+
+	mrs := rrq.Entries
+	t := btc.c.Open(rrq.TableName)
+
+	keys := make([]string, len(mrs))
+	muts := make([]*bigtable.Mutation, len(mrs))
+
+	for i, mr := range mrs {
+
+		key := string(mr.RowKey)
+		m := mutationFromProto(mr.Mutations)
+
+		// A little tricky here ... each key corresponds to a single Mutation
+		// object, where the indices of each slice must be sync'ed.
+		keys[i] = key
+		muts[i] = m
+	}
+
+	log.Printf("keys: %v", keys)
+	log.Printf("mutations: %v", muts)
+
+	ctx = btc.addCancelFunction(ctx)
+
+	errs, err := t.ApplyBulk(ctx, keys, muts)
+	if err != nil {
+		return nil, err
+	}
+	s.clientsLock.Unlock()
+
+	entries := make([]*btpb.MutateRowsResponse_Entry, 0)
+
+	// Iterate over any errors returned, matching indices with errors. If
+	// errs is nil, this block is skipped.
+	for i, e := range errs {
+		var me *btpb.MutateRowsResponse_Entry
+		if e != nil {
+			me = &btpb.MutateRowsResponse_Entry{
+				Index: int64(i),
+				Status: &status.Status{
+					Code:    int32(codes.Internal),
+					Message: e.Error(),
+				},
+			}
+
+		}
+		entries = append(entries, me)
```

Comment:
This will sometimes add nil to entries. Is that what you intend?

Reply:
Nope! Nice catch.